ansible/roles/test/tasks/pfc_wd/functional_test/functional_test.yml

```diff
@@ -0,0 +1,299 @@
+        run_cleanup: false
+      include: roles/test/tasks/run_config_test.yml
+
+    - set_fact:
+        test_expect_file: "expect_pfc_wd_detect"
+
+    - name: Initialize loganalyzer
+      include: roles/test/files/tools/loganalyzer/loganalyzer_init.yml
+
+    - name: Generate PFC storm on fanout switch
+      action: apswitch template=pfc_storm.j2
+      args:
+        host: "{{peer_device}}"
+        login: "{{peer_login}}"
+      connection: switch
+
+    - name: Check if logs contain message that PFC WD detected deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_analyze.yml
+
+    - name: Check if logs contain message that PFC WD detected deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_end.yml
+
+    - name: "Send 10K packets to {{pfc_wd_test_port}}"
+      include: roles/test/tasks/ptf_runner.yml
+      vars:
+        ptf_test_name: PFC WD test
+        ptf_test_dir: ptftests
+        ptf_test_path: pfc_wd.PfcWdTest
+        ptf_platform: remote
+        ptf_test_params:
+          - testbed_type='{{testbed_type}}'
+          - router_mac='{{ansible_Ethernet0['macaddress']}}'
+          - queue_index='{{pfc_queue_index}}'
+          - pkt_count='10000'
+          - port_src='{{pfc_wd_test_port_id}}'
+          - ip_src='{{pfc_wd_rx_neighbor_addr}}'
+          - ip_dst='{{pfc_wd_test_neighbor_addr}}'
+          - wd_action='forward'
+        ptf_extra_options: "--relax --debug info --log-file /tmp/pfc_wd.PfcWdTest.{{ansible_date_time.iso8601}}.log "
+
+    - set_fact:
+        test_expect_file: "expect_pfc_wd_restore"
+
+    - name: Initialize loganalyzer
+      include: roles/test/files/tools/loganalyzer/loganalyzer_init.yml
+
+    - name: Wait for queue to recover from PFC storm
+      pause:
+        seconds: "{{(pfc_wd_penalty / 1000 * 2) | int}}"
+
+    - name: Check if logs contain message that PFC WD restored from deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_analyze.yml
+
+    - name: Check if logs contain message that PFC WD restored from deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_end.yml
+
+  rescue:
+    - name: Remove all temporary files from local host.
+      file:
+        name: "{{ tests_location }}/functional_test/{{ item }}"
+        state: absent
+      with_items: "{{ config_files }}"
+      connection: local
+      become: false
+
+  always:
+    - name: Delete forward config from {{ pfc_wd_test_port }}.
+      vars:
+        config_file: pfc_wd_del_action.json
+        errors_expected: false
+        run_cleanup: false
+      include: roles/test/tasks/run_config_test.yml
+    - name: Remove all the temporary files created by the test.
+      include: roles/test/tasks/deinit_config_test.yml
+
+# Verify timers
+- name: Initialize config test
+  include: roles/test/tasks/init_config_test.yml
+
+- block:
+    - name: Apply forward config to {{ pfc_wd_test_port }}.
+      vars:
+        config_file: pfc_wd_fwd_action.json
```

Comment:
Can we apply drop action here? It is our key feature.

Reply:
both drop and fwd action should be verified.
if we have no other case for drop we should add so we have both options to be tested

Reply:
Sure. Agree to test both.

Reply:
It's a timer test. We don't even get to the actual action, but we verify the time difference between a moment when fanout starts sending PFC frames and when WD detects it. We don't even get to applying actions here. For the purpose of verifying that traffic is dropped / forwarded when a queue is mitigated from receiving PFC frames, there are separate tests above in this file.